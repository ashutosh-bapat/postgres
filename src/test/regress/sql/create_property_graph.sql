CREATE SCHEMA create_property_graph_tests;
SET search_path = create_property_graph_tests;

CREATE PROPERTY GRAPH g1;

COMMENT ON PROPERTY GRAPH g1 IS 'a graph';

CREATE PROPERTY GRAPH g1;  -- error: duplicate

CREATE TABLE t1 (a int, b text);
CREATE TABLE t2 (i int PRIMARY KEY, j int, k int);
CREATE TABLE t3 (x int, y text, z text);

CREATE TABLE e1 (a int, i int, t text, PRIMARY KEY (a, i));
CREATE TABLE e2 (a int, x int, t text);

CREATE PROPERTY GRAPH g2
    VERTEX TABLES (t1 KEY (a), t2, t3 KEY (x))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION t2,
        e2 KEY (a, x) SOURCE t1 DESTINATION t3
    );

-- error cases
CREATE PROPERTY GRAPH gx VERTEX TABLES (xx, yy);
CREATE PROPERTY GRAPH gx VERTEX TABLES (t1 KEY (a), t2 KEY (i), t1 KEY (a));
CREATE PROPERTY GRAPH gx
    VERTEX TABLES (t1 AS tt KEY (a), t2 KEY (i))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION t2
    );
CREATE PROPERTY GRAPH gx
    VERTEX TABLES (t1 KEY (a), t2 KEY (i))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION tx
    );
COMMENT ON PROPERTY GRAPH gx IS 'not a graph';

\dG

-- TODO
\d g1
\d+ g1

DROP TABLE g2;  -- error: wrong object type

DROP PROPERTY GRAPH g2;

DROP PROPERTY GRAPH g2;  -- error: does not exist

DROP PROPERTY GRAPH IF EXISTS g2;

DROP SCHEMA create_property_graph_tests CASCADE;
