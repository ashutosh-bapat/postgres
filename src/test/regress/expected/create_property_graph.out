CREATE SCHEMA create_property_graph_tests;
SET search_path = create_property_graph_tests;
CREATE PROPERTY GRAPH g1;
COMMENT ON PROPERTY GRAPH g1 IS 'a graph';
CREATE PROPERTY GRAPH g1;  -- error: duplicate
ERROR:  relation "g1" already exists
CREATE TABLE t1 (a int, b text);
CREATE TABLE t2 (i int PRIMARY KEY, j int, k int);
CREATE TABLE t3 (x int, y text, z text);
CREATE TABLE e1 (a int, i int, t text, PRIMARY KEY (a, i));
CREATE TABLE e2 (a int, x int, t text);
CREATE PROPERTY GRAPH g2
    VERTEX TABLES (t1 KEY (a), t2, t3 KEY (x))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION t2,
        e2 KEY (a, x) SOURCE t1 DESTINATION t3
    );
-- error cases
CREATE PROPERTY GRAPH gx VERTEX TABLES (xx, yy);
ERROR:  relation "xx" does not exist
CREATE PROPERTY GRAPH gx VERTEX TABLES (t1 KEY (a), t2 KEY (i), t1 KEY (a));
ERROR:  alias "t1" used more than once as element table
LINE 1: ...Y GRAPH gx VERTEX TABLES (t1 KEY (a), t2 KEY (i), t1 KEY (a)...
                                                             ^
CREATE PROPERTY GRAPH gx
    VERTEX TABLES (t1 AS tt KEY (a), t2 KEY (i))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION t2
    );
ERROR:  source vertex "t1" of edge "e1" does not exist
LINE 4:         e1 SOURCE t1 DESTINATION t2
                ^
CREATE PROPERTY GRAPH gx
    VERTEX TABLES (t1 KEY (a), t2 KEY (i))
    EDGE TABLES (
        e1 SOURCE t1 DESTINATION tx
    );
ERROR:  destination vertex "tx" of edge "e1" does not exist
LINE 4:         e1 SOURCE t1 DESTINATION tx
                ^
COMMENT ON PROPERTY GRAPH gx IS 'not a graph';
ERROR:  relation "gx" does not exist
-- information schema
SELECT * FROM information_schema.property_graphs ORDER BY property_graph_name;
 property_graph_catalog |    property_graph_schema    | property_graph_name 
------------------------+-----------------------------+---------------------
 regression             | create_property_graph_tests | g1
 regression             | create_property_graph_tests | g2
(2 rows)

SELECT * FROM information_schema.pg_element_tables ORDER BY property_graph_name, element_table_alias;
 property_graph_catalog |    property_graph_schema    | property_graph_name | element_table_alias | element_table_kind | table_catalog |        table_schema         | table_name | element_table_definition 
------------------------+-----------------------------+---------------------+---------------------+--------------------+---------------+-----------------------------+------------+--------------------------
 regression             | create_property_graph_tests | g2                  | e1                  | EDGE               | regression    | create_property_graph_tests | e1         | 
 regression             | create_property_graph_tests | g2                  | e2                  | EDGE               | regression    | create_property_graph_tests | e2         | 
 regression             | create_property_graph_tests | g2                  | t1                  | VERTEX             | regression    | create_property_graph_tests | t1         | 
 regression             | create_property_graph_tests | g2                  | t2                  | VERTEX             | regression    | create_property_graph_tests | t2         | 
 regression             | create_property_graph_tests | g2                  | t3                  | VERTEX             | regression    | create_property_graph_tests | t3         | 
(5 rows)

SELECT * FROM information_schema.pg_element_table_key_columns ORDER BY property_graph_name, element_table_alias, ordinal_position;
 property_graph_catalog |    property_graph_schema    | property_graph_name | element_table_alias | column_name | ordinal_position 
------------------------+-----------------------------+---------------------+---------------------+-------------+------------------
 regression             | create_property_graph_tests | g2                  | e1                  | a           |                1
 regression             | create_property_graph_tests | g2                  | e1                  | i           |                2
 regression             | create_property_graph_tests | g2                  | e2                  | a           |                1
 regression             | create_property_graph_tests | g2                  | e2                  | x           |                2
 regression             | create_property_graph_tests | g2                  | t1                  | a           |                1
 regression             | create_property_graph_tests | g2                  | t2                  | i           |                1
 regression             | create_property_graph_tests | g2                  | t3                  | x           |                1
(7 rows)

SELECT * FROM information_schema.pg_element_table_labels ORDER BY property_graph_name, element_table_alias, label_name;
 property_graph_catalog |    property_graph_schema    | property_graph_name | element_table_alias | label_name 
------------------------+-----------------------------+---------------------+---------------------+------------
 regression             | create_property_graph_tests | g2                  | e1                  | e1
 regression             | create_property_graph_tests | g2                  | e2                  | e2
 regression             | create_property_graph_tests | g2                  | t1                  | t1
 regression             | create_property_graph_tests | g2                  | t2                  | t2
 regression             | create_property_graph_tests | g2                  | t3                  | t3
(5 rows)

SELECT * FROM information_schema.pg_labels ORDER BY property_graph_name, label_name;
 property_graph_catalog |    property_graph_schema    | property_graph_name | label_name 
------------------------+-----------------------------+---------------------+------------
 regression             | create_property_graph_tests | g2                  | e1
 regression             | create_property_graph_tests | g2                  | e2
 regression             | create_property_graph_tests | g2                  | t1
 regression             | create_property_graph_tests | g2                  | t2
 regression             | create_property_graph_tests | g2                  | t3
(5 rows)

--\dG
-- TODO
\d g1
Property graph "create_property_graph_tests.g1"
 Column | Type 
--------+------

\d+ g1
Property graph "create_property_graph_tests.g1"
 Column | Type | Storage 
--------+------+---------
Property graph definition:
 CREATE PROPERTY GRAPH create_property_graph_tests.g1

DROP TABLE g2;  -- error: wrong object type
ERROR:  "g2" is not a table
HINT:  Use DROP PROPERTY GRAPH to remove a property graph.
DROP PROPERTY GRAPH g2;
DROP PROPERTY GRAPH g2;  -- error: does not exist
ERROR:  property graph "g2" does not exist
DROP PROPERTY GRAPH IF EXISTS g2;
NOTICE:  property graph "g2" does not exist, skipping
DROP SCHEMA create_property_graph_tests CASCADE;
NOTICE:  drop cascades to 6 other objects
DETAIL:  drop cascades to property graph g1
drop cascades to table t1
drop cascades to table t2
drop cascades to table t3
drop cascades to table e1
drop cascades to table e2
